### PR TITLE
Add new DisplayNameGenerator ReplaceCamelCaseAndUnderscoreAndNumber 

### PR DESCRIPTION
**Pull request by FanJups**
This extension handles method names with CamelCase, underscore and numbers.

The aim is to simplify unit test display names. Instead of using this method annotation org.junit.jupiter.api.DisplayName on every method, we can just use this class annotation org.junit.jupiter.api.DisplayNameGeneration and use that method annotation if needed.

 This generator follows 3 rules:

 - Each uppercase letter is turned into its lowercase value prepended by space.
 - Each underscore is turned into space. Words bounded by underscores or just starting with underscore are not transformed. Usually these words represent classes, variables...
 - Each number is prepended by space.

Closes: #793
PR: #819

---
**PR checklist**

The following checklist shall help the PR's author, the reviewers and maintainers to ensure the quality of this project.
It is based on our contributors guidelines, especially the ["writing code" section](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.adoc#writing-code).
It shall help to check for completion of the listed points.
If a point does not apply to the given PR's changes, the corresponding entry can be simply marked as done. 

Documentation (general)
* [x] There is documentation (Javadoc and site documentation; added or updated)
* [x] There is implementation information to describe _why_ a non-obvious source code / solution got implemented
* [x] Site documentation has its own `.adoc` file in the `docs` folder, e.g. `docs/report-entries.adoc`
* [x] Site documentation in `.adoc` file references demo in `src/demo/java` instead of containing code blocks as text
* [x] Only one sentence per line (especially in `.adoc` files)
* [x] Javadoc uses formal style, while sites documentation may use informal style

Documentation (new extension)
* [x] The `docs/docs-nav.yml` navigation has an entry for the new extension
* [x] The `package-info.java` contains information about the new extension

Code (general)
* [x] Code adheres to code style, naming conventions etc.
* [x] Successful tests cover all changes
* [x] There are checks which validate correct / false usage / configuration of a functionality and there are tests to verify those checks
* [x] Tests use [AssertJ](https://assertj.github.io/doc/) or our own [PioneerAssert](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.adoc#assertions) (which are based on AssertJ)

Code (new package)
* [x] The new package is exported in `module-info.java`
* [x] The new package is also present in the tests
* [x] The new package is opened for reflection to JUnit 5 in `module-info.java`
* [x] The new package is listed in the contribution guide

Contributing
* [x] A prepared commit message exists
* [x] The list of contributions inside `README.adoc` mentions the new contribution (real name optional)

### CODE REVIEW

**Comment by Michael1993** [comment]
Personally, I'd prefer to review this when the build is passing - if you need help, let us know.

**Comment by FanJups** [comment]
> Personally, I'd prefer to review this when the build is passing - if you need help, let us know.

Sure, thanks.  One of my test is failing.  The goal is to use this generator as a 

> Personally, I'd prefer to review this when the build is passing - if you need help, let us know.

Sure, thanks.  I am trying to test @DisplayNameGeneration(ReplaceCamelCaseAndUnderscoreAndNumber.class) but my test failed. 

That's why it's still in draft mode.

**Comment by FanJups** [comment]
> > Personally, I'd prefer to review this when the build is passing - if you need help, let us know.
> 
> Sure, thanks. I am trying to test @DisplayNameGeneration(ReplaceCamelCaseAndUnderscoreAndNumber.class) but my test failed.
> 
> That's why it's still in draft mode.

Maybe, its  because the generator is not well configured.

**Comment by FanJups** [comment]
Hi @Michael1993 , please are available these days for a call ?

**Comment by FanJups** [comment]
@Michael1993 , please I need help on code formatting. The feature and docs re ready

**Comment by Michael1993** [comment]
> @Michael1993 , please I need help on code formatting. The feature and docs re ready

You can just run the `spotlessApply` task, it will automatically format the code for you.

**Comment by FanJups** [comment]
> > @Michael1993 , please I need help on code formatting. The feature and docs re ready
> 
> You can just run the `spotlessApply` task, it will automatically format the code for you.

Thanks for the tip, it's done.

**Comment by FanJups** [comment]
@Michael1993 , I just corrected  javadoc, please run the workflow

**Comment by FanJups** [comment]
I am trying to figure out why I got these  [javadoc errors](https://github.com/junit-pioneer/junit-pioneer/actions/runs/9963278306/job/27529089637?pr=819)   here but not on my IDE ? 🤔

**Comment by FanJups** [comment]
> I am trying to figure out why I got these [javadoc errors](https://github.com/junit-pioneer/junit-pioneer/actions/runs/9963278306/job/27529089637?pr=819) here but not on my IDE ? 🤔

Got this article trying to explain the matter  https://reflectoring.io/howto-format-code-snippets-in-javadoc/

> <pre> + {@code}
> Combining <pre> and {@code}, indentations and line breaks are kept and '<' and '>' don’t have to be escaped. However, against all expectations the '@' character is now evaluated as a Javadoc tag. What’s worse: it cannot even be escaped using the HTML number code, since the HTML number code would be literalized by {@code}.

**Comment by FanJups** [comment]
> > I am trying to figure out why I got these [javadoc errors](https://github.com/junit-pioneer/junit-pioneer/actions/runs/9963278306/job/27529089637?pr=819) here but not on my IDE ? 🤔
> 
> Got this article trying to explain the matter https://reflectoring.io/howto-format-code-snippets-in-javadoc/
> 
> > pre  + {@code}
> > Combining pre and {@code}, indentations and line breaks are kept and '<' and '>' don’t have to be escaped. However, against all expectations the '@' character is now evaluated as a Javadoc tag. What’s worse: it cannot even be escaped using the HTML number code, since the HTML number code would be literalized by {@code}.

Using &#064; does not render as @ on my ide

**Comment by FanJups** [comment]
> > > I am trying to figure out why I got these [javadoc errors](https://github.com/junit-pioneer/junit-pioneer/actions/runs/9963278306/job/27529089637?pr=819) here but not on my IDE ? 🤔
> > 
> > 
> > Got this article trying to explain the matter https://reflectoring.io/howto-format-code-snippets-in-javadoc/
> > > pre  + {@code}
> > > Combining pre and {@code}, indentations and line breaks are kept and '<' and '>' don’t have to be escaped. However, against all expectations the '@' character is now evaluated as a Javadoc tag. What’s worse: it cannot even be escaped using the HTML number code, since the HTML number code would be literalized by {@code}.
> 
> Using @ does not render as @ on my ide

A bug ? https://bugs.openjdk.org/browse/JDK-8130754

**Comment by Michael1993** [comment]
Try using `<pre>` tags, instead of `{@code}`.

**Comment by FanJups** [comment]
> Try using `\<pre\>` tags, instead of `{@code}`.

font, indentations and line breaks will be lost

**Comment by FanJups** [comment]
https://bugs.openjdk.org/browse/JDK-8130754

```
/home/runner/work/junit-pioneer/junit-pioneer/src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java:36: error: unterminated inline tag
  \<pre\> {@code
```

**Comment by FanJups** [comment]
> This illustrates whats going on https://bugs.openjdk.org/browse/JDK-8130754
> 
> @code works well with @ when its on the same line
> 
> ```
> /home/runner/work/junit-pioneer/junit-pioneer/src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java:36: error: unterminated inline tag
>   \<pre\> {@code
> ```

@Michael1993 , I updated following this bug page, please run again

**Comment by FanJups** [comment]
There was a format issue due to my last change, now I believe everything will be ok. @Michael1993

**Comment by FanJups** [comment]
perfect 😊, thanks @Michael1993 for your review.

File: src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java
Comment (Michael1993) [inline src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java:22]: Usage example here is unnecessary - javadoc is for technical details, IMO.

File: src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java
Comment (Michael1993) [inline src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java:26]: I would like to highlight this section as the perfect example of something that is both important for a user documentation and the javadoc.

File: src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java
Comment (Michael1993) [inline src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java:34]: Personally, I am a big fan of code examples in javadoc, so kudos for that. That being said, I don't think this is particularly useful here... I would rather see some configuration examples instead.

File: src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java
Comment (Michael1993) [inline src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java:152]: I am not outright against your approach, however I think I'd prefer something a bit more concise. What do you think of my approach?
```suggestion
	private String replaceCamelCaseAndUnderscoreAndNumber(String input) {
		// Remove leading underscore(s)
		var sanitized = input.replaceAll("^_+(.*)$", "$1");
		List<String> list = new ArrayList<>();
		String[] split = sanitized.split("_");
		for (int i = 0; i < split.length; i++) {
			if (i % 2 == 1) {
				// If parity is odd, we are between two underscores, no formatting necessary
				list.add(split[i]);
			} else {
				list.add(formatCamelCase(split[i]));
			}
		}
		// Some cases can lead to double spaces - i.e.: underscore (closing) followed by capital letter
		var joined = String.join(" ", list).replaceAll("\\s{2}", " ");
		// Capitalize the first letter
		return joined.substring(0, 1).toUpperCase() + joined.substring(1);
	}

	private static String formatCamelCase(String in) {
		return in.replaceAll("(\\d+)", " $1").replaceAll("([A-Z]+)", " $1").toLowerCase();
	}
```

File: src/test/java/org/junitpioneer/jupiter/displaynamegenerator/MethodTestDescriptor.java
Comment (Michael1993) [inline src/test/java/org/junitpioneer/jupiter/displaynamegenerator/MethodTestDescriptor.java:23]: These are unused?

**Review by Michael1993 — CHANGES_REQUESTED** [review]
Exceptional work @FanJups ! Thank you for your diligent contribution. I do have some small comments.

File: src/main/java/module-info.java
Comment (FanJups) [reply src/main/java/module-info.java:27]: I created another package because this is a new purpose. I wanted the package name to be as clear as possible to the developer in order to easily guess the link with JUnit 5 DisplayNameGenerator.  Also, I thought there will be other DisplayName generators in the future. Furthermore,  testing this new feature requires to create other classes then grouping these test classes in the same package is a good idea.

**Comment by FanJups** [comment]
> Exceptional work @FanJups ! Thank you for your diligent contribution. I do have some small comments.

Thanks to you, I was able to implement this idea 😊. Thanks for the opportunity.

File: docs/docs-nav.yml
Comment (FanJups) [reply docs/docs-nav.yml]: ok, just updated.

File: src/test/java/org/junitpioneer/jupiter/displaynamegenerator/MethodTestDescriptor.java
Comment (FanJups) [reply src/test/java/org/junitpioneer/jupiter/displaynamegenerator/MethodTestDescriptor.java:23]: Yes, I will delete that

File: docs/replace-camelcase-and-underscore-and-number.adoc
Comment (FanJups) [reply docs/replace-camelcase-and-underscore-and-number.adoc]: Updated

File: src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java
Comment (FanJups) [reply src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java:22]: updated

File: src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java
Comment (FanJups) [reply src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java:26]: updated

File: src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java
Comment (FanJups) [reply src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java:34]: updated

File: src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java
Comment (FanJups) [reply src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java]: updated

**Comment by FanJups** [comment]
Hi @Michael1993 , please any updates ?

**Comment by Michael1993** [comment]
> Hi @Michael1993 , please any updates ?

Hi @FanJups - terribly sorry, my work got a little hectic. I think the code is fine as-is but the project has a two-maintainer policy, so I am waiting for another maintainer (maybe @beatngu13 ?) to approve and merge this.

**Comment by FanJups** [comment]
> > Hi @Michael1993 , please any updates ?
> 
> Hi @FanJups - terribly sorry, my work got a little hectic. I think the code is fine as-is but the project has a two-maintainer policy, so I am waiting for another maintainer (maybe @beatngu13 ?) to approve and merge this.

No worries,  same on my side. It's just that I wanted your feedback. Thanks again for this opportunity 😊

**Comment by Bukama** [comment]
> > Hi @Michael1993 , please any updates ?
> 
> Hi @FanJups - terribly sorry, my work got a little hectic. I think the code is fine as-is but the project has a two-maintainer policy, so I am waiting for another maintainer (maybe @beatngu13 ?) to approve and merge this.

And I was waiting for you to finish yours so I see your comments were already in place :D I guess we call this deadlock? :D

File: src/test/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumberTestEngine.java
Comment (beatngu13) [inline src/test/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumberTestEngine.java:59]: I'm not familiar with `TestEngine` API, but according to the [Javadoc](https://junit.org/junit5/docs/current/api/org.junit.platform.engine/org/junit/platform/engine/TestEngine.html#getArtifactId()):

>Get the Artifact ID of the JAR in which this test engine is packaged. 

That's not the artifact ID? And why do we need a separate engine, for testing?

Comment (FanJups) [reply src/test/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumberTestEngine.java:59]: It's because the one used by JUnit to test their display name generators is internal . In order to test this feature,  I get inspired by what they did

https://github.com/junit-team/junit5/blob/main/jupiter-tests/src/test/java/org/junit/jupiter/api/DisplayNameGenerationTests.java

Comment (FanJups) [reply src/test/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumberTestEngine.java:59]: This doc is talking about the default implementation .
This article helped me a lot

https://blogs.oracle.com/javamagazine/post/junit-build-custom-test-engines-java

Comment (FanJups) [reply src/test/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumberTestEngine.java:59]: > It's because the one used by JUnit to test their display name generators is internal . In order to test this feature, I get inspired by what they did
> 
> https://github.com/junit-team/junit5/blob/main/jupiter-tests/src/test/java/org/junit/jupiter/api/DisplayNameGenerationTests.java

I tried to extend the class **AbstractJupiterTestEngineTests** but it was not possible,  that's why I concluded that I have to implement my own Test Engine.
It was my first time to deal with Test engine.  While figuring out how to do that, I got this article from Oracle

https://blogs.oracle.com/javamagazine/post/junit-build-custom-test-engines-java

File: README.adoc
Comment (Bukama) [reply README.adoc:126]: > As far as I remember, we (Nicolai) are in the [Oxford Comma](https://www.grammarly.com/blog/punctuation-capitalization/what-is-the-oxford-comma-and-why-do-people-care-so-much-about-it/) camp. 😉 (Same goes for docs and elsewhere.)

Yes we are (or the English is, but well)

File: src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java
Comment (FanJups) [reply src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java:25]: Looks like see tag does not recognize such relative path then I added 

`<a href="https://junit-pioneer.org/docs/replace-camelcase-and-underscore-and-number">Usage example of ReplaceCamelCaseAndUnderscoreAndNumber</a>`

File: src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java
Comment (FanJups) [reply src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java:34]: Looks like see tag does not recognize such relative path then I added

`<a href="https://junit-pioneer.org/docs/replace-camelcase-and-underscore-and-number">Usage example of ReplaceCamelCaseAndUnderscoreAndNumber</a>`

File: src/test/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumberTestEngine.java
Comment (beatngu13) [reply src/test/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumberTestEngine.java:59]: Ah, I see. Thanks for the great explanation and those helpful links! 👍

File: src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java
Comment (beatngu13) [reply src/main/java/org/junitpioneer/jupiter/displaynamegenerator/ReplaceCamelCaseAndUnderscoreAndNumber.java:25]: I'm sorry, I meant the same comments count as in the AsciiDoc file. But that's fine, I will resolve the conversation(s).

**Review by beatngu13 — APPROVED** [review]

**Comment by beatngu13** [comment]
@FanJups thanks a lot, great work! I will merge the PR as soon as the builds are passing.

**Pull request merged**
